public/app/features/dashboard/components/ShareModal/SharePublicDashboard.tsx

```diff
@@ -1,60 +1,85 @@
 import { dispatch } from 'app/store/store';
 
 import {
   dashboardCanBePublic,
   getPublicDashboardConfig,
   savePublicDashboardConfig,
+  getDashboard,
   PublicDashboardConfig,
+  DashboardResponse,
 } from './SharePublicDashboardUtils';
 import { ShareModalTabProps } from './types';
 
 interface Props extends ShareModalTabProps {}
 
-// 1. write test for dashboardCanBePublic
-// 2. figure out how to disable the switch
-
 export const SharePublicDashboard = (props: Props) => {
-  const [publicDashboardConfig, setPublicDashboardConfig] = useState<PublicDashboardConfig>({ isPublic: false });
   const dashboardUid = props.dashboard.uid;
+  const orgId = contextSrv.user.orgId;
+  const [publicDashboardConfig, setPublicDashboardConfig] = useState<PublicDashboardConfig>({
+    isPublic: false,
+    publicDashboard: { uid: '', dashboardUid, orgId },
+  });
+
+  const [dashboard, setDashboard] = useState<DashboardModel>({} as DashboardModel);
 
   useEffect(() => {
+    // dashboard model may be stale, so load it ourselves
+    getDashboard(dashboardUid)
+      .then((dashboardResponse: DashboardResponse) => {
+        setDashboard(new DashboardModel(dashboardResponse.dashboard, dashboardResponse.meta));
+      })
+      .catch((err) => {
+        dispatch(notifyApp(createErrorNotification('Failed to retrieve dashboard', err)));
+      });
+
+    // load config
     getPublicDashboardConfig(dashboardUid)
       .then((pdc: PublicDashboardConfig) => {
         setPublicDashboardConfig(pdc);
       })
-      .catch(() => {
-        dispatch(notifyApp(createErrorNotification('Failed to retrieve public dashboard config')));
+      .catch((err) => {
+        dispatch(notifyApp(createErrorNotification('Failed to retrieve public dashboard config', err)));
       });
   }, [dashboardUid]);
 
   const onSavePublicConfig = () => {
-    // verify dashboard can be public
-    if (!dashboardCanBePublic(props.dashboard)) {
+    if (!dashboardCanBePublic(dashboard)) {
       dispatch(notifyApp(createErrorNotification('This dashboard cannot be made public')));
       return;
     }
 
-    try {
-      savePublicDashboardConfig(props.dashboard.uid, publicDashboardConfig);
-      dispatch(notifyApp(createSuccessNotification('Dashboard sharing configuration saved')));
-    } catch (err) {
-      console.error('Error while making dashboard public', err);
-      dispatch(notifyApp(createErrorNotification('Error making dashboard public')));
-    }
+    savePublicDashboardConfig(props.dashboard.uid, publicDashboardConfig)
+      .then((pdc: PublicDashboardConfig) => {
+        setPublicDashboardConfig(pdc);
+        dispatch(notifyApp(createSuccessNotification('Dashboard sharing configuration saved')));
+      })
+      .catch((err) => {
+        console.error('Error while making dashboard public', err);
+        dispatch(notifyApp(createErrorNotification('Error making dashboard public')));
+      });
```

Comment:
Can we move the error & notification handling to the utils file (to keep the react component more focused on rendering, and not processing actions). And try to use async await instead of Promise directly 

>

Reply:
Done